app/src/main/java/org/groundplatform/android/ui/util/DistanceFormatExt.kt

```diff
@@ -0,0 +1,108 @@
+ * limitations under the License.
+ */
+package org.groundplatform.android.ui.util
+
+import android.content.res.Resources
+import android.icu.text.MeasureFormat
+import android.icu.util.LocaleData
+import android.icu.util.Measure
+import android.icu.util.MeasureUnit
+import android.icu.util.ULocale
+import android.os.Build
+import java.util.Locale
+import kotlin.math.roundToInt
+import org.groundplatform.android.R
+import timber.log.Timber
+
+private const val METERS_TO_FEET = 3.28084
+private const val DISTANCE_THRESHOLD = 10.0
+private const val DECIMAL_MULTIPLIER = 100.0
+
+fun Double.formatDistance(resources: Resources): String? {
+  if (this <= 0) return null
+
+  val locale = Locale.getDefault()
+  val isImperial = isImperialSystem(locale)
+  val (convertedDistance, unitLabel, measureUnit) = this.convertDistance(resources, isImperial)
+  val roundedNumber = roundDistanceAsDouble(convertedDistance)
+  val fallback = formatDistanceFallback(roundedNumber, unitLabel, locale)
+
+  return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+    formatWithMeasureFormat(roundedNumber, measureUnit, fallback, locale)
+  } else {
+    fallback
+  }
+}
+
+private fun Double.convertDistance(
+  resources: Resources,
+  isImperial: Boolean,
+): Triple<Double, String, MeasureUnit> =
+  if (isImperial) {
+    Triple(this * METERS_TO_FEET, resources.getString(R.string.unit_feet), MeasureUnit.FOOT)
+  } else {
+    Triple(this, resources.getString(R.string.unit_meters), MeasureUnit.METER)
+  }
+
+private fun roundDistanceAsDouble(distance: Double): Double =
+  if (distance < DISTANCE_THRESHOLD) {
+    (distance * DECIMAL_MULTIPLIER).roundToInt() / DECIMAL_MULTIPLIER
+  } else {
+    distance.roundToInt().toDouble()
+  }
+
+private fun formatDistanceFallback(number: Double, unitLabel: String, locale: Locale): String {
+  val formatted =
+    if (number < 10) {
+      String.format(locale, "%.2f", number)
+    } else {
+      String.format(locale, "%.0f", number)
+    }
+  return "$formatted $unitLabel"
+}
+
+private fun formatWithMeasureFormat(
+  number: Double,
+  measureUnit: MeasureUnit,
+  fallback: String,
+  locale: Locale,
+): String =
+  runCatching {
+      val formatter =
+        MeasureFormat.getInstance(ULocale.forLocale(locale), MeasureFormat.FormatWidth.SHORT)
+      formatter.format(Measure(number, measureUnit))
+    }
+    .getOrElse { exception ->
+      Timber.w("FormatDistance", "MeasureFormat failed", exception)
+      fallback
+    }
+
+private fun isImperialSystem(locale: Locale): Boolean =
+  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+    runCatching {
+        LocaleData.getMeasurementSystem(ULocale.forLocale(locale)) ==
+          LocaleData.MeasurementSystem.US
+      }
+      .getOrElse { exception ->
+        Timber.w("MeasurementSystem", "Failed to get measurement system from LocaleData", exception)
+        isImperialSystemFallback(locale)
+      }
+  } else {
+    isImperialSystemFallback(locale)
```

Comment:
Can we short circuit to reduce nesting here, checking and returning this value first on older Androids?